### ISSUE TITLE
Typemap errors even when theres no errors

### ISSUE THREAD

**Issue by aalhitennf** [serenity-rs/serenity#2301]
I dont completely understand what is happening, and why. My bot is very similar to basic examples from repo, nothing special to show, and this is how my usual context loading from typemap looks like:  

```
#[command]
#[description = "Do whatever"]
async fn whatever(ctx: &Context, msg: &Message) -> CommandResult {

    let data_read = ctx.data.read().await;

    let whatever_db = data_read.get::<MongoDBCollection<IWhatever>>().ok_or({
         log::error!("No whatever database context");
         CommandError::from("No whatever database context")
    })?;

     let xd = whatever_db.find("whatever")... and so on...
    ...
```

So, this code works, but somehow still outputs this error message "No whatever database context". What am i doing wrong here or is this something that serenity does?

**Comment by arqunis** [comment]
Would you mind posting your `TypeMapKey` implementation for `MongoDBCollection<T>`, and the code where you specifically insert `MongoDBCollection<IWhatever>` to the typemap?

**Comment by aalhitennf** [comment]
TypeMapKey implementations
```
impl<T: Send + Sync + 'static> TypeMapKey for MongoDBCollection<T> {
    type Value = MongoDBCollection<T>;
}


impl TypeMapKey for Counter {
    type Value = HashMap<u64, Instant>;
}
```

Insert
```
...
let whatever1: MongoDBCollection<IWhatever1> = MongoDBCollection::new(some, args, here);
let whatever2: MongoDBCollection<IWhatever2> = MongoDBCollection::new(some, args, here);
let whatever3: MongoDBCollection<IWhatever3> = MongoDBCollection::new(some, args, here);
let whatever4: MongoDBCollection<IWhatever4> = MongoDBCollection::new(some, args, here);

 let builder = Client::builder(token, intents)
        .event_handler(Handler::new())
        .framework(framework)
        .type_map_insert::<Counter>(HashMap::default())
        .type_map_insert::<MongoDBCollection<IWhatEver1>>(whatever1)
        .type_map_insert::<MongoDBCollection<IWhatEver2>>(whatever2)
        .type_map_insert::<MongoDBCollection<IWhatever3>>(whatever3)
        .type_map_insert::<MongoDBCollection<IWhatever4>>(whatever4);
```  

Errors shows up when accessing Counter, these collections and whatever else i insert into the typemap. But yet it still works without problems, im confused.

**Comment by arqunis** [comment]
> But yet it still works without problems, im confused.

Ah, I thought you had set up the typemap incorrectly, but it's fine. The problem here isn't even related to typemap, but the language -- specifically, eagerness.

When you call a function using the `name(arg1, arg2, arg3...)` syntax, the language will first evaluate each argument before proceeding to calling the function. This is the default mode in many languages, called "eager evaluation". That is, a value is resolved before it's actually used. For example:
```rust
let x = 4 + 2;
...
let y = x * 4;
```
where `x` and `y` are distantly spaced out. The expression `4 + 2` will be evaluated straight away upon reaching the `let x = ...;` line, despite the first use of `x` being in the expression of `y`. In most cases, this is the desired behaviour, but sometimes it's not, especially if instead of, say, `2` in `4 + 2`, we had a call to `frobnicate()` (`4 + frobnicate()`) that was really expensive to compute (it has to make a network request or something). In this instance, it would be nice to prolongate the evaluation of `x` until `y`.

This is where "lazy evaluation" is useful, and in a language like Haskell is the default. In Rust, this can be emulated with a closure:
```rust
let x = || 4 + frobnicate();
...
let y = x() * 4;
```

The closure will get evaluated, but not its body, which we can run later whenever we want.

So in short, what you want instead of `ok_or` is [`ok_or_else`](https://doc.rust-lang.org/nightly/std/option/enum.Option.html#method.ok_or_else). What you are currently using is a block, which is just a series of statements. It's as if you wrote:

```rust
#[command]
#[description = "Do whatever"]
async fn whatever(ctx: &Context, msg: &Message) -> CommandResult {

    let data_read = ctx.data.read().await;

    log::error!("No whatever database context");
    let whatever_db = data_read.get::<MongoDBCollection<IWhatever>>()
         .ok_or(CommandError::from("No whatever database context"))?;

     let xd = whatever_db.find("whatever")... and so on...
    ...

**Comment by aalhitennf** [comment]
Oh i see, that explains why clippy often nags about using ok_or and recommends ok_or_else, i just never did understand why. Thanks for explaining that, now i know 🤝
